packages/async/src/babel-plugin.ts

```diff
@@ -0,0 +1,143 @@
+import * as Types from '@babel/types';
+import {NodePath, Binding} from '@babel/traverse';
+
+const DEFAULT_PROCESS_PACKAGES = {
+  '@shopify/react-async': ['createAsyncComponent'],
+  '@shopify/react-graphql': ['createAsyncQueryComponent'],
+};
+
+interface State {
+  processPackages?: Map<string, string[]>;
+  opts?: {packages?: {[key: string]: string[]}};
+}
+
+export default function asyncBabelPlugin({types: t}: {types: typeof Types}) {
+  return {
+    visitor: {
+      Program(_path: NodePath<Types.Program>, state: State) {
+        state.processPackages = new Map(
+          Object.entries(
+            (state.opts && state.opts.packages) || DEFAULT_PROCESS_PACKAGES,
+          ),
+        );
+      },
+      ImportDeclaration(path: NodePath<Types.ImportDeclaration>, state: State) {
+        const {processPackages} = state;
+
+        if (!(processPackages instanceof Map)) {
+          return;
+        }
+
+        const source = path.node.source.value;
+
+        const processImports = processPackages.get(source) || [];
+
+        if (processImports.length === 0) {
+          return;
+        }
+
+        const factoryImports = path.get('specifiers').filter(specifier => {
+          return (
+            specifier.isImportSpecifier() &&
+            processImports.some(name =>
+              specifier.get('imported').isIdentifier({name}),
+            )
+          );
+        });
+
+        if (factoryImports.length === 0) {
+          return;
+        }
+
+        for (const factoryImport of factoryImports) {
+          const bindingName = factoryImport.node.local.name;
+          const binding = path.scope.getBinding(bindingName);
+          if (binding != null) {
+            processBinding(binding, t);
+          }
+        }
+      },
+    },
+  };
+}
+
+function processBinding(binding: Binding, t: typeof Types) {
```

Comment:
what is a binding? How are we processing it?

Reply:
called it `addIdOption`, but `Binding` is also a common term and has a type definition right next to it.